### DISCUSSION TITLE
Support for processes running inside docker containers

### DISCUSSION THREAD

**Discussion by pierricgimmig** [google/orbit#4532] [Q&A]
It seems there was some work done in this direction in #4195, but it was reverted with #4257. Is there any plan to resume work on this @danielfenner ? Thanks!

**Comment by pierricgimmig** [comment]
Why was the change reverted by the way?

**Comment by florian-kuebler** [comment]
Providing a quick answer for now. @danielfenner will likely sharing a document about the details publicly soon.

So #4195 was adding the tid translation for user-space dynamic instrumentation, but Vulkan layer and manual instrumentation was missing at that point. The biggest issue was AFAIR, "where to do the processing", as we require all events to be in order. For user-space instrumentation, we pipe the events back to the perf_event_open events, and maintain ordering here.
For manual instrumentation and the vulkan layer, reordering happens at a later point. Fixing the tid's would require a more serious refactoring of the event pipeline in OrbitService.

Another minor thing that I remember vaguely is, that for we apparently did not had all required information for the events available (I think some were missing process id?).

In the end the requirement for tid namespaces became obsolete, and as #4195 was not giving a full-fledged solution, we decided to revert.

Also, we internally came to the conclusion that probably, the less error-prone approach is to run OrbitService inside the namespace. Is that also a solution in your case? If not, please wait until we publish the document about our efforts.

**Reply by danielfenner** [reply]
Hey Pierric, hey Florian, 

sorry for ignoring this for so long - I never looked into the discussions and so this fell through the cracks. Florian just made me aware of this thread. 

I'll move some of the write-up to the wiki so we can decide where to go with it. 

tldr: Florian wrote most of the above already. The solution in https://github.com/google/orbit/pull/4195 is incomplete and I also was not happy about the performance. At the time of reverting it I thought deploying the service into the container might be the solution but this comes with other problems. Most importantly permissions e.g. for perf_event_open might be missing from inside the container. So maybe https://github.com/google/orbit/pull/4195 was the right way to go after all.

**Comment by danielfenner** [comment]
I just added a bunch of thoughts here: https://github.com/google/orbit/wiki/Make-Orbit-aware-of-pid-namespaces

As mentioned above the performance issue with the solution from https://github.com/google/orbit/pull/4195 is the main blocker here. Thinking about this a bit more I think that in principle the solution to translate all the pid's to root namespace pid's is a good decision. The alternative is to deploy into the namespace of the target process but this comes with a bunch of problems and as well (there is a bit about this in the above write-up).
